drivers/flash/spi_nor.c

```diff
@@ -0,0 +1,522 @@
+/*
+ * Copyright (c) 2018 Savoir-Faire Linux.
+ *
+ * This driver is heavily inspired from the spi_flash_w25qxxdv.c SPI NOR driver.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <flash.h>
+#include <spi.h>
+#include <init.h>
+#include <string.h>
+#include "spi_nor.h"
+#include "flash_priv.h"
+
+#define SZ_256  0x100
+#define SZ_512  0x200
+#define SZ_1024 0x400
+#define SZ_4K   0x1000
+#define SZ_32K  0x8000
+#define SZ_64K  0x10000
+
+#define MASK_256 0xFF
+#define MASK_4K  0xFFF
+#define MASK_32K 0x7FFF
+#define MASK_64K 0xFFFF
+
+#define BLOCK_ERASE_32K BIT(1)
+#define BLOCK_ERASE_64K BIT(2)
+
+#define SPI_NOR_MAX_ID_LEN 3
+#define SPI_NOR_MAX_ADDR_WIDTH 4
+
+#define JEDEC_ID(x)		    \
+	{			    \
+		((x) >> 16) & 0xFF, \
+		((x) >> 8) & 0xFF,  \
+		(x) & 0xFF,	    \
+	}
+
+/**
+ * struct flash_info - Structure for defining the SPI NOR layer
+ * @name: Name of the flash
+ * @id: Jedec ID
+ * @page_size: Size of the pages
+ * @sector_size: Size of the sectors
+ * @n_sectors: Number of sector
+ * @block_size: Size of the blocks
+ * @flag: Flags of the flash
+ */
+static const struct flash_info {
+	char *name;
+	u8_t id[SPI_NOR_MAX_ID_LEN];
+	u32_t page_size;
+	u32_t sector_size;
+	u32_t n_sectors;
+	u32_t block_size;
+	u32_t flag;
+} flash_ids[] = {
+	{ CONFIG_SPI_NOR_DRV_NAME, JEDEC_ID(CONFIG_SPI_NOR_JEDEC_ID),
+	CONFIG_SPI_NOR_PAGE_SIZE, CONFIG_SPI_NOR_SECTOR_SIZE,
+	CONFIG_SPI_NOR_SECTORS, CONFIG_SPI_NOR_BLOCK_SIZE,
+	CONFIG_SPI_NOR_BLOCK_ERASE_64K},
+};
```

Comment:
All this information can passed as a device configuration struct that to ```DEVICE_AND_API_INIT```, in this case there is no need anymore for ```char* name```.